packages/fxa-payments-server/src/lib/oauth.ts

```diff
@@ -0,0 +1,230 @@
+} = {}) {
+  // Stash our current location to resume after authorization.
+  storage.setItem(storageKey(clientId, 'resumeLocation'), window.location.href);
+
+  // Generate the code verifier for PKCE
+  const verifier = generateRandomCode();
+  storage.setItem(storageKey(clientId, 'verifier'), verifier);
+
+  // Generate a random state for return by FxA after authorization as a
+  // CSRF-prevention mechanism
+  // https://tools.ietf.org/html/rfc6749#section-10.12
+  const state = generateRandomCode();
+  storage.setItem(storageKey(clientId, 'state'), state);
+
+  const params: { [propName: string]: string } = {
+    client_id: clientId,
+    response_type: 'code',
+    redirect_uri: `${window.location.origin}/`,
+    code_challenge: await hashVerifier(verifier),
+    code_challenge_method: 'S256',
+    scope: scopes.join(' '),
+    state,
+  };
+
+  if (promptNone && loginHint) {
+    // We can use ?prompt=none only if we also have a login hint
+    params.prompt = 'none';
+    params.login_hint = loginHint;
+  }
+
+  const query = new URLSearchParams(Object.entries(params));
+  window.location.replace(`${authorizationEndpoint}?${query.toString()}`);
+}
+
+export type FetchTokenWithCodeResult = {
+  access_token: string;
+  token_type: string;
+  scope: string;
+  expires_in: number;
+  auth_at: number;
+};
+
+export async function fetchTokenWithCode({
+  tokenEndpoint = `${config.servers.auth.url}/v1/token`,
+  code = '',
+  state = '',
+  clientId = config.servers.oauth.clientId,
+  storage = window.localStorage,
+} = {}): Promise<FetchTokenWithCodeResult> {
+  const verifier = storage.getItem(storageKey(clientId, 'verifier'));
+
+  // FxA promises that the state query param should match the state we
+  // passed at the start of authorization as a CSRF-prevention mechanism
+  // https://tools.ietf.org/html/rfc6749#section-10.12
+  const storedState = storage.getItem(storageKey(clientId, 'state'));
+  if (state !== storedState) {
+    throw new Error(
+      `state stored at start of authorization does not match state returned at end ${state} !== ${storedState}`
+    );
+  }
+
+  const response = await fetch(tokenEndpoint, {
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    method: 'POST',
+    body: JSON.stringify({
+      client_id: clientId,
+      code,
+      grant_type: 'authorization_code',
+      code_verifier: verifier,
+    }),
+  });
+
+  if (!response.ok) {
+    throw new Error(`fetchTokenWithCode failed ${response.status}`);
+  }
+
+  return await response.json();
+}
+
+const storageKey = (clientId: string, name: string) =>
+  `oauth-authorization-${encodeURIComponent(clientId)}-${name}`;
+
+function generateRandomCode() {
+  return Array.from(window.crypto.getRandomValues(new Uint32Array(16)))
+    .map(c => c.toString(16))
+    .join('');
+}
+
+export async function hashVerifier(verifier: string) {
+  // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
```

Comment:
Using [WebCrypto](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) here - we can use a node crypto polyfill here, but it looks like that more than doubles the overall size of our app bundle. 😓  WebCrypto seems like it's been around awhile in browser support. Maybe we don't need a polyfill?

Reply:
FWIW our existing https://github.com/mozilla/fxa-pairing-channel already assumes that WebCrypto will be available and fails ungracefully if it's not. You could consider throwing a specific error here if it's missing, to give metrics on whether and how much it's actually missing in practice.